text/0317-third-party-dependency-management.md

```diff
@@ -0,0 +1,129 @@
+
+1. Quickly reacting when a security vulnerability is discovered in a version of a dependency used by the project.
+
+2. Protecting ourselves against [supply chain attacks](https://en.wikipedia.org/wiki/Supply_chain_attack),
+  like [typosquatting](https://www.darkreading.com/vulnerabilities---threats/beware-the-package-typosquatting-supply-chain-attack/a/d-id/1340383),
+  malicious package takeovers, etc.
+
+## Current situation
+
+### Dependencies
+
+Right now, our released packages bundle all of their dependencies.
+
+#### CLIs
+
+In our CLI packages (`aws-cdk`, `cdk-assets`, etc.) that are NodeJS-only,
+we have a very large dependency closure
+(around 180 packages - [see here](https://npmgraph.js.org/?q=aws-cdk)
+for the details).
+
+At build time of these packages,
+we generate an `npm-shrinkwrap.json` file with the exact version of each dependency
+(including transitive ones).
+This means a customer installing those packages in a given version will always get all of their dependencies in a specific version
+(the one defined in `npm-shrinkwrap.json`) as well.
+
+#### Framework
+
+In our framework packages, we have a much smaller dependency closure,
+roughly equal to:
+
+- `yaml@1.10.2`, which has 0 dependencies
+- `semver@7.3.5`, which depends on `lru-cache@6.0.0`, which depends on `yalist@4.0.0`
+- `jsonschema@1.4.0` (from `@aws-cdk/cloudformation-schema`)
+- `punycode@2.1.1` (from `@aws-cdk/aws-cognito`)
+- `case@1.6.3` (from `@aws-cdk/aws-codepipeline-actions`)
+- `@aws-cdk/core` depends on several packages for bundling: `ignore@5.1.8`, `@balena/dockerignore@1.0.2`, and:
+  - `minimatch@3.0.4`, which depends on `brace-expansion@1.1.11`, which depends on `balanced-match@1.02`, `concat-map@0.0.1`
+  - `fs-extra@9.1.0`, which depends on `at-least-node@1.0.0`, `jsonfile@6.1.0`, which depends on `universalify@2.0.0`, `graceful-fs@4.2.6`
+
+All of these third-party dependencies are bundled with the released framework packages.
+This is a JSII requirement
+(otherwise, these 3rd-party dependencies would not be available when using CDK from other languages!).
+
+For framework libraries in non-Node languages,
+the only extra dependencies
+(in addition to the bundled Node ones)
+are generally only the appropriate JSII runtime.
+These are updated separately, and are easier for customers to manage,
+as non-Node package managers make it easy to override the version of a transitive dependency
+like the JSII runtime in case a security vulnerability is found in it.
+Because of that, this RFC only focuses on managing Node dependencies.
+
+### Updating
+
+We have a GitHub Action that once a week updates all dependencies in the CDK repository to their latest,
+usually minor, version,
+using the [`npm-check-updates`](https://www.npmjs.com/package/npm-check-updates) package.
+
+### Summary
+
+Our current situation defends us pretty well against supply-chain attacks,
+because of the bundling that we do
```

Comment:
Only true for framework dependencies.

And also managing transitive dependencies is pretty much a pain...

Reply:
This is not trivial. The dependency chain is huge. The cli is the right place to start imho. And this would set examples for other CDKs. Its definitely going involve some refactors.

Reply:
Replaced "audit" with "investigate" in the doc, and called out that it might affect some CLI functionality.